url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/beans/Statement.java

```diff
@@ -0,0 +1,610 @@
+import java.security.PrivilegedAction;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+import org.apache.harmony.beans.BeansUtils;
+import org.apache.harmony.beans.internal.nls.Messages;
+
+public class Statement {
+
+    private Object target;
+
+    private String methodName;
+
+    private Object[] arguments;
+
+    // cache used methods of specified target class to accelerate method search
+    private static WeakHashMap<Class<?>, Method[]> classMethodsCache = new WeakHashMap<Class<?>, Method[]>();
+
+    public Statement(Object target, String methodName, Object[] arguments) {
+        this.target = target;
+        this.methodName = methodName;
+        this.arguments = arguments == null ? BeansUtils.EMPTY_OBJECT_ARRAY
+                : arguments;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        if (target == null) {
+            sb.append(BeansUtils.NULL);
+        } else {
+            Class<?> clazz = target.getClass();
+            sb.append(clazz == String.class ? BeansUtils.QUOTE : BeansUtils
+                    .idOfClass(clazz));
+        }
+        sb.append('.' + methodName + '(');
+        if (arguments != null) {
+            Class<?> clazz;
+            for (int index = 0; index < arguments.length; index++) {
+                if (index > 0) {
+                    sb.append(", "); //$NON-NLS-1$
+                }
+                if (arguments[index] == null) {
+                    sb.append(BeansUtils.NULL);
+                } else {
+                    clazz = arguments[index].getClass();
+                    sb.append(clazz == String.class ? '"' + (String) arguments[index] + '"'
+                            : BeansUtils.idOfClass(clazz));
+                }
+            }
+        }
+        sb.append(')');
+        sb.append(';');
+        return sb.toString();
+    }
+
+    public String getMethodName() {
+        return methodName;
+    }
+
+    public Object[] getArguments() {
+        return arguments;
+    }
+
+    public Object getTarget() {
+        return target;
+    }
+
+    public void execute() throws Exception {
+        invokeMethod();
+    }
+
+    Object invokeMethod() throws Exception {
+        Object result = null;
+        try {
+            Object target = getTarget();
+            String methodName = getMethodName();
+            Object[] arguments = getArguments();
+            Class<?> targetClass = target.getClass();
+            if (targetClass.isArray()) {
+                Method method = findArrayMethod(methodName, arguments);
+                Object[] copy = new Object[arguments.length + 1];
+                copy[0] = target;
+                System.arraycopy(arguments, 0, copy, 1, arguments.length);
+                result = method.invoke(null, copy);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)